### ISSUE TITLE
[Feature Request] Linear constraints between candidates in a q-batch in optimize_acqf

### ISSUE THREAD

**Issue by jduerholt** [pytorch/botorch#1737]
# 🚀 Feature Request

## Motivation

Imagine a system in which you can run several experiments in parallel, but all experiments have to have to share the same value for at least one input parameter. Let's make it more concrete:
 
A system can run several chemical reactions in parallel, but one can only set one temperature in the system which then holds for all minireactors in the system, whereas all the other degrees of freedom can be set independently between the different reactors/candidates. It is related to a fixed feature but it is not the same as a fixed feature. The temperature can be proposed by the BO, but vor every candidate in the proposed q-batch it has to be the same temperature. How to find under this constraint the maximum acquisition function? 

## Pitch

From my understanding, this could be implemented as a linear equality constraint between the different candidates in the q-batch and the SLSQP should be able to handle it. If you request for example 2 candidates, it should hold `T_1 == T_2` which is the same as `T_1 - T_2 ==0`, where `T` is the temperature.

Any ideas, how to add these kind of inter-candidate linear constraints to `optimize_acqf`?

**Are you willing to open a pull request?** Yes.

**Comment by saitcakmak** [comment]
That's an interesting application. I think you should be able to do this in a similar way to how we construct the constraints for scipy optimizer. I am not super familiar with that part of the code, but the existing setup duplicates the simple constraints for each batch dimension, so that they're applied to all candidates. You'd instead want to construct the constraints to span the q-batch dimension and only duplicate them for the t-batch dimensions. I think this should be relatively straightforward as long as you don't use `sequential=True` for optimization.

This is where the constraints are constructed: https://github.com/pytorch/botorch/blob/main/botorch/optim/parameter_constraints.py#L68

cc @esantorella who might be more familiar with how this all works

**Comment by saitcakmak** [comment]
Actually, it looks like there's some support for constraints across q-batch dimensions: https://github.com/pytorch/botorch/blob/main/botorch/optim/parameter_constraints.py#L290-L301

**Comment by Balandat** [comment]
Indeed this should already be supported that way (lmk if you're running into any problems). 

cc @Ryan-Rhys @qingfeng10 - this is the mechanism to enforce e.g. a budget constraint a batch of `q` candidates.

**Comment by jduerholt** [comment]
Wow, I am very impressed. It works out of the box. Here is an example snippet:

```python
import torch
from botorch.models import SingleTaskGP
from botorch.fit import fit_gpytorch_mll
from botorch.utils import standardize
from gpytorch.mlls import ExactMarginalLogLikelihood
from botorch.acquisition import UpperConfidenceBound, qUpperConfidenceBound
from botorch.optim import optimize_acqf
tkwargs = {'dtype': torch.float64, 'device': 'cpu'}

train_X = torch.rand(10, 2).to(**tkwargs)
Y = 1 - torch.norm(train_X - 0.5, dim=-1, keepdim=True)
Y = Y + 0.1 * torch.randn_like(Y)  # add some noise
train_Y = standardize(Y)

gp = SingleTaskGP(train_X, train_Y)
mll = ExactMarginalLogLikelihood(gp.likelihood, gp)
fit_gpytorch_mll(mll)

UCB = qUpperConfidenceBound(gp, beta=0.1)

constraint = [(torch.tensor([[0,0], [1,0]]), torch.tensor([1.0, -1.0]), 0), (torch.tensor([[0,0], [2,0]]), torch.tensor([1.0, -1.0]), 0)]
batch_initial_conditions = torch.rand(5,3,2).to(**tkwargs)


bounds = torch.stack([torch.zeros(2), torch.ones(2)])
candidate, acq_value = optimize_acqf(
    UCB, bounds=bounds, q=3, num_restarts=5, raw_samples=20, equality_constraints=constraint, batch_initial_conditions=batch_initial_conditions
)
candidate  

```

It returns for example:

```
tensor([[0.4165, 0.1331],
        [0.4165, 0.4348],
        [0.4165, 0.3929]], dtype=torch.float64)
```

Testing this, a few things catched my attention:

- One should document this possibility also in the docstring of `optimize_acqf`. Else it is relatively difficult for users to find this option.
- If you set it up without `batch_initial_conditions` it crahes with a message from the polytope sampling process (which is clear), but I think we should catch this before and tell the user that one has to use `batch_initial_conditions` in this case. On the long run one could think about adding these constraints also to the samplers.

What do you think?

**Comment by saitcakmak** [comment]
Happy to see it supports your use case! I think both are good suggestions. We can probably add the validation in [`OptimizeAcqfInputs.__post_init__`](https://github.com/pytorch/botorch/blob/main/botorch/optim/optimize.py#L92). As always, happy to accept a PR for this :)

**Comment by jduerholt** [comment]
I further thought about the problem, and this kind of constraint is already implemented in the hit and run polytope sampler, we just have to switch how we think about a sample from the polytope sampler. Currently the polytope sampler is used to generate samples sample by sample. For example if we have a 2dimensional problem and we want to generate a q-batch of 3, one does the following:

``` python
from botorch.utils.sampling import get_polytope_samples
import torch
tkwargs = {'dtype': torch.float64, 'device': 'cpu'}

get_polytope_samples(
    n=3,
    bounds=torch.tensor([[0,0], [1,1]]).to(**tkwargs),
    inequality_constraints=[(torch.tensor([0,1], dtype=torch.int64), torch.tensor([-1,-1]).to(**tkwargs), -1)]
)
```

This returns a tensor of shape `3x2`. To generate 5 q-batches, we just do this 5 times, but of course only the "horizontal" constraint will be fulfilled.

In the scenario where we have vertical q-batch constraints we have to change perspective. The sampler should now return as one sample the complete q-batch. This can be achieved by:

``` python
from botorch.utils.sampling import get_polytope_samples
import torch
tkwargs = {'dtype': torch.float64, 'device': 'cpu'}

q = 3
num_restarts = 5
dim = 2
lower = [0]*dim
upper = [1]*dim

# normal constraints
constraints = [(torch.tensor([i*dim, i*2+1], dtype=torch.int64), torch.tensor([-1,-1]).to(**tkwargs), -1) for i in range(q)]
# q constraints
equalities = [(torch.tensor([0, 2*i], dtype=torch.int64), torch.tensor([1,-1]).to(**tkwargs), 0) for i in range(1, q)]

samples = get_polytope_samples(
    n=num_restarts,
    bounds=torch.tensor([lower*q, upper*q]).to(**tkwargs),
    inequality_constraints=constraints,
    equality_constraints = equalities)

samples.reshape(torch.Size((num_restarts,q,dim)))
```
This will return a tensor of shape `5x3x2` in which both horizontal and vertical constraints are fulfilled like here:

```
tensor([[[0.0177, 0.0845],
         [0.0177, 0.9182],
         [0.0177, 0.4823]],

        [[0.0664, 0.0801],
         [0.0664, 0.5995],
         [0.0664, 0.6444]],

        [[0.0144, 0.7214],
         [0.0144, 0.4225],
         [0.0144, 0.7300]],

        [[0.0083, 0.6065],
         [0.0083, 0.5582],
         [0.0083, 0.4035]],

        [[0.0814, 0.8063],
         [0.0814, 0.7706],
         [0.0814, 0.8157]]], dtype=torch.float64)
```

From this follows that we have to change the generation of the intial conditions in a way that if these q-batch constraints are present the problem for the polytope sampler is transformed in this way, and then we have found a general way to deal with it. 

Any thoughts on this, especially what to implement where?

**Comment by Balandat** [comment]
> This will return a tensor of shape `5x3x2` in which both horizontal and vertical constraints are fulfilled like here:

What exactly are the vertical constraints here (sorry having a bit of a hard time parsing this from the code)? I assume by "horizontal" constraints you mean that the first parameter be the same across all points of the q-batch. Not sure what you mean the "vertical" constraints here though.

**Comment by jduerholt** [comment]
Horizontal constraints are the usual constraints which just affect the candidates in one q batch (rows) independent from each other. Here that the sum of them is smaller equal to one. 

With vertical constraints, I mean the constraints across candidates within a q batch (the ones we are mainly discussing in this issue). Here that all candidates/rows in one q batch have the same value for the first value. The problem is that the polytope sampler cannot handle them currently. In the code above, I show how to transform the problem in a way that they can be handled by the polytope sampler. Implementing this will allow to give both type of constraints to optimize_acqf without the need to specify batch_initial_conditions. 

I hope this makes it more clearly:-)

**Comment by Balandat** [comment]
Yup, got it, thanks. This makes sense to me; instead of sampling on a space `P` you're sampling on the product space `P x ... x P` with the additional constraint across the elements of the q-batch.

**Comment by Balandat** [comment]
@jduerholt this should be resolved by now with #1757, right?

**Comment by jduerholt** [comment]
Yes, it is resolved!
